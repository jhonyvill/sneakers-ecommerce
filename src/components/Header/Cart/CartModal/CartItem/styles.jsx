import styled from "styled-components";

export const Item = styled.div`
  display: grid;
  grid-template-columns: auto 1fr auto;
  grid-template-rows: 1fr 1fr;
  align-items: center;
  gap: 0 0.625rem;
  max-height: 3.125rem;
`;

export const Photo = styled.img`
  max-width: 3.125rem;
  grid-column: 1;
  grid-row: span 2;
  border-radius: 4px;
`;

export const Title = styled.h2`
  font-size: inherit;
  font-weight: 400;
  grid-column: 2;
  grid-row: 1;
  margin: 0;
  max-width: 30ch;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
`;

export const PriceContainer = styled.div`
  display: flex;
  gap: 0.5rem;
  grid-column: 2;
  grid-row: 2;
  font-weight: 400;
`;

export const TotalPrice = styled.span`
  font-weight: 700;
  color: ${({ theme }) => theme.colors.blackOpacity};
`;

export const DeleteButton = styled.button`
  background-color: transparent;
  grid-column: 3;
  grid-row: span 2;
  padding: 0;

  &:hover svg > * {
    fill: ${({ theme }) => theme.colors.primaryOrange};
  }
`;
