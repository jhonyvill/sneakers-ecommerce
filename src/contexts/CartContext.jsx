import React from "react";
import { useReducer } from "react";

export const CartContext = React.createContext();

export const CartProvider = ({ children }) => {
  const initialState = { items: [] };

  function cartReducer(state, action) {
    switch (action.type) {
      case "ADD_ITEM":
        return { ...state, items: [...state.items, action.payload] };
      case "UPDATE_ITEM":
        const { product } = action.payload;
        const itemIndex = state.items.findIndex(
          (item) => item.product.id === product.id
        );

        const updatedItems = [...state.items];
        updatedItems.slice(itemIndex, product);
        return { ...state, items: updatedItems };
      default:
        state;
    }
  }

  const [state, dispatch] = useReducer(cartReducer, initialState);

  return (
    <>
      <CartContext.Provider
        value={{ cartState: state, cartDispatch: dispatch }}
      >
        {children}
      </CartContext.Provider>
    </>
  );
};
