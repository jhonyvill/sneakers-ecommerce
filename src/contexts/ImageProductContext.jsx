import React from "react";
import { useState } from "react";

export const ImageProductContext = React.createContext();

export const ImageProductProvider = ({ children }) => {
  const [selectedImage, setSelectedImage] = useState({
    image: null,
    index: 0,
  });

  const handleSelectedModalImage = (image, index) => {
    setSelectedImage({ image, index });
  };

  return (
    <>
      <ImageProductContext.Provider
        value={{
          imageSelected: selectedImage.image,
          indexImageSelected: selectedImage.index,
          handleSelectedModalImage,
        }}
      >
        {children}
      </ImageProductContext.Provider>
    </>
  );
};
