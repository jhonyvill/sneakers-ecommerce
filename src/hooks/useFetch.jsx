import { useCallback, useState } from "react";

export const useFetch = () => {
  const [data, setData] = useState(null);
  const [error, setError] = useState(null);
  const [loading, setLoading] = useState(false);

  const request = useCallback(async (url) => {
    let response;
    let json;
    try {
      setError(null);
      setLoading(true);

      response = await fetch(url);
      json = await response.json();
      if (!response.ok) throw new Error(json.message);
    } catch (error) {
      json = null;
      setError(error.message);
      console.log(error.message);
    } finally {
      setData(json);
      setLoading(false);
    }
  }, []);

  return {
    data,
    error,
    loading,
    request,
  };
};
