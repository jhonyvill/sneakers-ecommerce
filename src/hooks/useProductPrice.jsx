import React from "react";

export const useProductPrice = (product) => {
  const discountPercent = product.discount * 100;
  const price = product.price.toFixed(2);
  const discountValue = price * product.discount;
  const finalPrice = (price - discountValue).toFixed(2);
  const hasDiscount = discountPercent > 0;

  function getTotalItemPrice(quantity) {
    return (finalPrice * quantity).toFixed(2);
  }

  return { price, finalPrice, discountPercent, hasDiscount, getTotalItemPrice };
};
