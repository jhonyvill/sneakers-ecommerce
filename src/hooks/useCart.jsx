import React from "react";
import { useState, useContext } from "react";
import { CartContext } from "../contexts/CartContext";

export const useCart = () => {
  const { cartState, cartDispatch } = useContext(CartContext);
  const [quantitySelected, setQuantitySelected] = useState(0);
  const cartHasItems = cartState.items.length > 0;

  function findProductInCart(product) {
    return cartState.items.find((item) => item.product.id === product.id);
  }

  function addCartProduct(product) {
    const existentProduct = findProductInCart(product);
    if(existentProduct){
      cartDispatch({
        type: "UPDATE_ITEM",
        payload: {
          product: product,
          quantity: (existentProduct.quantity += quantitySelected),
        },
      });
    }else if(quantitySelected > 0){
      cartDispatch({
        type: "ADD_ITEM",
        payload: { product: product, quantity: quantitySelected },
      });
    }
    setQuantitySelected(0);
  }

  return {
    quantitySelected,
    setQuantitySelected,
    addCartProduct,
    cartHasItems
  };
};
