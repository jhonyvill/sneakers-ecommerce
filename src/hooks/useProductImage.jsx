import React from "react";
import { useContext } from "react";
import { useState } from "react";
import { ImageProductContext } from "../contexts/ImageProductContext";

export const useProductImage = (images) => {
  const { imageSelected, indexImageSelected, handleSelectedModalImage } = useContext(ImageProductContext);
  const [featuredImage, setFeaturedImage] = useState({
    image: imageSelected || images[0],
    index: indexImageSelected,
  });

  function handleChangeImage(image, indexClicked) {
    indexClicked >= 0 &&
      indexClicked <= images.length - 1 &&
      indexClicked !== featuredImage.index &&
      setFeaturedImage({ image, index: indexClicked });
  }
  function handleClickFeaturedImage() {
    handleSelectedModalImage(featuredImage.image, featuredImage.index);
  }

  return {
    featuredImage: featuredImage.image,
    imageIndexActive: featuredImage.index,
    handleChangeImage,
    handleClickFeaturedImage,
  };
};
