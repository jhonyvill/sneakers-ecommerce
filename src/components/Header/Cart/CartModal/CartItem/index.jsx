import React from "react";
import * as S from "./styles";
import TrashSvg from "../../../../../assets/icon-delete.svg?react";
import { useProductPrice } from "../../../../../hooks/useProductPrice";
import { useCart } from "../../../../../hooks/useCart";
import PropTypes from "prop-types";

const CartItem = ({ item }) => {
  const { finalPrice, getTotalItemPrice } = useProductPrice(item.product);
  const totalItemPrice = getTotalItemPrice(item.quantity);
  const { removeCartProduct } = useCart();

  return (
    <S.Item>
      <S.Photo
        src={item.product.pictures[0].thumb}
        alt={item.product.pictures[0].alt}
        width="176px"
        height="176px"
      />
      <S.Title>{item.product.title}</S.Title>
      <S.PriceContainer>
        <span>{`$${finalPrice} x ${item.quantity}`}</span>
        <S.TotalPrice>${totalItemPrice}</S.TotalPrice>
      </S.PriceContainer>
      <S.DeleteButton
        onClick={() => removeCartProduct(item.product.id)}
        aria-label="deletar item"
      >
        <TrashSvg />
      </S.DeleteButton>
    </S.Item>
  );
};

CartItem.propTypes = {
  item: PropTypes.object,
};

export default CartItem;
