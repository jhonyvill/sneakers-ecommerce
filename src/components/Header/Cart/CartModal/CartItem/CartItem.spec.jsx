import { screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { renderWithTheme, itemDefault } from "../../../../../../.jest/utils";
import * as hookCart from "../../../../../hooks/useCart";
import { CartProvider } from "../../../../../contexts/CartContext";
import CartItem from "./";

function renderCartItem(item = itemDefault) {
  return renderWithTheme(
    <CartProvider>
      <CartItem item={item} />
    </CartProvider>
  );
}

const item = { ...itemDefault, quantity: 2 };

function createUseCartSpy(mockFunction) {
  return jest
    .spyOn(hookCart, "useCart")
    .mockImplementationOnce(() => ({ removeCartProduct: mockFunction }));
}

describe("<CartItem />", () => {
  it("should render CartItem component", () => {
    const { getByTestId } = renderCartItem(item);
    const totalPrice = getByTestId("total-price");

    expect(totalPrice).toBeInTheDocument();
    expect(totalPrice).toHaveTextContent("$250.00");
    expect(screen.getByText("$125.00 x 2")).toBeInTheDocument();
    expect(screen.getByRole("img", { name: /product 1/i })).toBeInTheDocument();
    expect(screen.getByRole("heading", { name: /title mock/i })).toBeInTheDocument();
    expect(screen.getByRole("button", { name: /deletar item/i })).toBeInTheDocument();
    expect(getByTestId("trash-svg")).toBeInTheDocument();
  });

  it("should call removeCartProduct function when delete button clicked", async () => {
    const mockFunction = jest.fn();
    createUseCartSpy(mockFunction);
    renderCartItem(item);
    const deleteButton = screen.getByRole("button", { name: /deletar item/i });

    expect.assertions(3);
    expect(deleteButton).toBeInTheDocument();

    await userEvent.click(deleteButton);

    expect(mockFunction).toHaveBeenCalledTimes(1);
    expect(mockFunction).toHaveBeenCalledWith(itemDefault.product.id);
  });
});

describe("<CartItem /> snapshots", () => {
  it("should match snapshot", () => {
    const {container} = renderCartItem(item);
    expect(container.firstChild).toMatchSnapshot();
  });
});
