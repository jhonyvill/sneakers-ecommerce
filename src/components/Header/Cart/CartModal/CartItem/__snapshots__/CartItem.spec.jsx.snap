// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CartItem /> snapshots should match snapshot 1`] = `
<div
  class="styles__Item-sc-1eal0cm-0 bCjpZ"
>
  <img
    alt="product 1"
    class="styles__Photo-sc-1eal0cm-1 cZzmFX"
    height="176px"
    src="thumbnail-1.jpg"
    width="176px"
  />
  <h2
    class="styles__Title-sc-1eal0cm-2 dOavNX"
  >
    title mock
  </h2>
  <div
    class="styles__PriceContainer-sc-1eal0cm-3 dGMGIu"
  >
    <span>
      $125.00 x 2
    </span>
    <span
      class="styles__TotalPrice-sc-1eal0cm-4 hJISjn"
      data-testid="total-price"
    >
      $
      250.00
    </span>
  </div>
  <button
    aria-label="deletar item"
    class="styles__DeleteButton-sc-1eal0cm-5 fCBTaV"
  >
    <icon-mock
      data-testid="trash-svg"
    />
  </button>
</div>
`;
